src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -989,82 +946,153 @@ HeapWord* G1CollectedHeap::satisfy_failed_allocation_helper(size_t word_size,
   // incremental pauses.  Therefore, at least for now, we'll favor
   // expansion over collection.  (This might change in the future if we can
   // do something smarter than full collection to satisfy a failed alloc.)
-  result = expand_and_allocate(word_size);
-  if (result != nullptr) {
-    return result;
-  }
+  if (expand(word_size)) {
+    return attempt_allocation_at_safepoint(word_size,
+                                           node_index,
+                                           expect_null_mutator_alloc_region);
 
-  if (do_gc) {
-    GCCauseSetter compaction(this, GCCause::_g1_compaction_pause);
-    // Expansion didn't work, we'll try to do a Full GC.
-    // If maximal_compaction is set we clear all soft references and don't
-    // allow any dead wood to be left on the heap.
-    if (maximal_compaction) {
-      log_info(gc, ergo)("Attempting maximal full compaction clearing soft references");
-    } else {
-      log_info(gc, ergo)("Attempting full compaction");
-    }
-    *gc_succeeded = do_full_collection(maximal_compaction /* clear_all_soft_refs */ ,
-                                       maximal_compaction /* do_maximal_compaction */);
   }
-
   return nullptr;
 }
 
-HeapWord* G1CollectedHeap::satisfy_failed_allocation(size_t word_size,
-                                                     bool* succeeded) {
+bool G1CollectedHeap::satisfy_failed_allocations(bool* gc_succeeded) {
   assert_at_safepoint_on_vm_thread();
 
-  // Attempts to allocate followed by Full GC.
-  HeapWord* result =
-    satisfy_failed_allocation_helper(word_size,
-                                     true,  /* do_gc */
-                                     false, /* maximum_collection */
-                                     false, /* expect_null_mutator_alloc_region */
-                                     succeeded);
+  bool success = handle_allocation_requests(false /* expect_null_mutator_alloc_region*/);
```

Comment:
Maybe it would be good to rename `success` to something more specific to avoid confusion with `gc_succeeded`. Something like `alloc_succeeded` (compared to succeeding the gc and succeeding the entire method).